File: core/src/com/ariasaproject/advancerofrpg/physics/box2d/Joint.java
package com.ariasaproject.advancerofrpg.physics.box2d;

import com.ariasaproject.advancerofrpg.math.Vector2;
import com.ariasaproject.advancerofrpg.physics.box2d.JointDef.JointType;

public abstract class Joint {
	protected long addr;
	private final World world;
	private final float[] tmp = new float[2];
	private Object userData;
	protected JointEdge jointEdgeA;
	protected JointEdge jointEdgeB;

	private static native void initialize();

	static {
		initialize();
	}

	protected Joint(World world, long addr) {
		this.world = world;
		this.addr = addr;
	}

	public JointType getType() {
		int type = jniGetType();
		if (type > 0 && type < JointType.valueTypes.length)
			return JointType.valueTypes[type];
		else
			return JointType.Unknown;
	}

	private native int jniGetType();

	public Body getBodyA() {
		return world.bodies.get(jniGetBodyA());
	}

	private native long jniGetBodyA();

	public Body getBodyB() {
		return world.bodies.get(jniGetBodyB());
	}

	private native long jniGetBodyB();

	private final Vector2 anchorA = new Vector2();

	public Vector2 getAnchorA() {
		jniGetAnchorA(tmp);
		anchorA.x = tmp[0];
		anchorA.y = tmp[1];
		return anchorA;
	}

	private native void jniGetAnchorA(float[] anchorA);

	private final Vector2 anchorB = new Vector2();

	public Vector2 getAnchorB() {
		jniGetAnchorB(tmp);
		anchorB.x = tmp[0];
		anchorB.y = tmp[1];
		return anchorB;
	}

	private native void jniGetAnchorB(float[] anchorB);

	public native boolean getCollideConnected();

	private final Vector2 reactionForce = new Vector2();

	public Vector2 getReactionForce(float inv_dt) {
		jniGetReactionForce(inv_dt, tmp);
		reactionForce.x = tmp[0];
		reactionForce.y = tmp[1];
		return reactionForce;
	}

	private native void jniGetReactionForce(float inv_dt, float[] reactionForce);

	public native float getReactionTorque(float inv_dt);

// /// Get the next joint the world joint list.
// b2Joint* GetNext();
//
	public Object getUserData() {
		return userData;
	}

	public void setUserData(Object userData) {
		this.userData = userData;
	}

	public native boolean isActive();

}

File: native-core/src/box2d/Joint.cpp
#include "Joint.h"
#include "Dynamics/Joints/b2Joint.h"

//java value id
static jfieldID jointPtr;

Joint_M(void, initialize)(JNIEnv *env, jclass clazz) {
    //initialize field ID of JNI
    jointPtr = env->GetFieldID(clazz, "addr", "J");
}

Joint_M(jint, jniGetType)(JNIEnv *env, jobject obj) {
    return ((b2Joint *) env->GetLongField(obj, jointPtr))->GetType();
}

Joint_M(jlong, jniGetBodyA)(JNIEnv *env, jobject obj) {
    return (jlong) ((b2Joint *) env->GetLongField(obj, jointPtr))->GetBodyA();
}

Joint_M(jlong, jniGetBodyB)(JNIEnv *env, jobject obj) {
    return (jlong) ((b2Joint *) env->GetLongField(obj, jointPtr))->GetBodyB();
}

Joint_M(void, jniGetAnchorA)(JNIEnv *env, jobject obj, jfloatArray vals) {
    jfloat *data = (jfloat *) env->GetPrimitiveArrayCritical(vals, 0);
    const b2Vec2 a = ((b2Joint *) env->GetLongField(obj, jointPtr))->GetAnchorA();
    data[0] = a.x;
    data[1] = a.y;
    env->ReleasePrimitiveArrayCritical(vals, data, 0);
}

Joint_M(void, jniGetAnchorB)(JNIEnv *env, jobject obj, jfloatArray vals) {
    jfloat *data = (jfloat *) env->GetPrimitiveArrayCritical(vals, 0);
    const b2Vec2 a = ((b2Joint *) env->GetLongField(obj, jointPtr))->GetAnchorB();
    data[0] = a.x;
    data[1] = a.y;
    env->ReleasePrimitiveArrayCritical(vals, data, 0);
}

Joint_M(jboolean, getCollideConnected)(JNIEnv *env, jobject obj) {
    return ((b2Joint *) env->GetLongField(obj, jointPtr))->GetCollideConnected();
}

Joint_M(void, jniGetReactionForce)(JNIEnv *env, jobject obj, jfloat inv_dt, jfloatArray vals) {
    jfloat *data = (jfloat *) env->GetPrimitiveArrayCritical(vals, 0);
    const b2Vec2 a = ((b2Joint *) env->GetLongField(obj, jointPtr))->GetReactionForce(inv_dt);
    data[0] = a.x;
    data[1] = a.y;
    env->ReleasePrimitiveArrayCritical(vals, data, 0);
}

Joint_M(jfloat, getReactionTorque)(JNIEnv *env, jobject obj, jfloat inv_dt) {
    return ((b2Joint *) env->GetLongField(obj, jointPtr))->GetReactionTorque(inv_dt);
}

Joint_M(jboolean, isActive)(JNIEnv *env, jobject obj) {
    return ((b2Joint *) env->GetLongField(obj, jointPtr))->IsActive();
}